### DISCUSSION TITLE
How does `deno_core` get notified when there is a new task in `queueMicroktask`?

### DISCUSSION THREAD

**Discussion by linrongbin16** [denoland/deno#25930] [Q&A] [answered]
Hi, I'm learning the `rusty_v8` and the microtask queue related APIs.

Does `deno_core` need to get notified when there's a new tasks in `queueMicrotask` ? It seems there's no way to get notified when a microtask is been triggered:

The APIs are:
- [`Isolate::enqueue_microtask`](https://docs.rs/v8/latest/v8/struct.Isolate.html#method.enqueue_microtask)
- [`MicrotaskQueue::is_running_microtasks`](https://docs.rs/v8/latest/v8/struct.MicrotaskQueue.html#method.is_running_microtasks) or [`Isolate::has_pending_background_tasks`](https://docs.rs/v8/latest/v8/struct.Isolate.html#method.has_pending_background_tasks)
- [`MicrotaskQueue::perform_checkpoint`](https://docs.rs/v8/latest/v8/struct.MicrotaskQueue.html#method.perform_checkpoint) or [`Isolate::perform_microtask_checkpoint`](https://docs.rs/v8/latest/v8/struct.Isolate.html#method.perform_microtask_checkpoint)

This means we don't actually have a way to _**block**_ and wait for the next microtask, instead we will have to do a forever loop. Here's a stupid sample code to show what I mean:

```rust
pub fn tick_event_loop(&mut self) {
  if self.isolate.has_pending_background_tasks() {
    self.isoalte.perform_microtask_checkpoint();
  }
}

pub fn run_event_loop(&mut self) {
  loop {
    self.tick_event_loop();
  }
}
```

I mean, in a js runtime such like node/deno, the event loop will _**block**_ the main thread and wait for callbacks trigger it, not forever loop and use 100% CPU core.

How does `deno_core` solve this issue?

I have some guesses:

1. `deno_core` implement the `queueMicrotask` API by itself. i.e. it creates a queue inside rust side, and expose an `ops` API (for example `ops.queue_microtask`) to javascript side, then set `globalThis.queueMicrotask = function(callback)  { ops.queue_microtask(callback);  }`. So `deno_core` can get notified when the queue is triggered.
2. The micro tasks never need to trigger the event loop to run (which should be a standard behavior specified in ECMAScript), it always simply wait to run in the next tick triggered by other `Promise`.

Would you please help me with it?

**Comment by bartlomieju** [comment, marked as answer]
> I mean, in a js runtime such like node/deno, the event loop will block the main thread and wait for callbacks trigger it, not forever loop and use 100% CPU core.

That's correct - only if there's some scheduled work the event loop is blocking. If there is no scheduled work then nothing can push more microtasks onto the queue.

> deno_core implement the queueMicrotask API by itself. i.e. it creates a queue inside rust side, and expose an ops API (for example ops.queue_microtask) to javascript side, then set globalThis.queueMicrotask = function(callback)  { ops.queue_microtask(callback);  }. So deno_core can get notified when the queue is triggered.

Not really, when microtask is queued it uses the `Isolate::enqueue_microtask` API and V8 handles them internally in its queue.

> The micro tasks never need to trigger the event loop to run (which should be a standard behavior specified in ECMAScript), it always simply wait to run in the next tick triggered by other Promise.

Correct. There must be some other scheduled work that can queue microtasks and potentially schedule more work for the future.

**Reply by linrongbin16** [reply]
really thanks for your teaching!
